README.md
#IBEN-FML
A fork of [IBEN 1.1](https://sourceforge.net/projects/iben/) originally created by Gerd Behrmann, with patches allowing compilation on modern Linux distributions, as well as other small improvements such as the ability to choose another postscript viewer then Ghostview as the standard version of IBEN 1.1 used.

## Dependencies
The following dependencies should be installed before compiling IBEN-FML.

+ [BuDDy](https://sourceforge.net/projects/buddy/)
+ A postscript compatible viewer, GV, GSview, etc.
+ graphviz 

## Installation
1. Install BuDDy, graphviz, and a postscript viewer, GSview is used as the default.

2. If another viewer then GSview is installed, then change the macro PSV in iben.h.
```
#define PSV "gsview" => #define PSV "your viewer"
```

3. Use the standard ./configure and make combination to compile the program.

## Patches
This is a list of all the changes made to the program, none of patches changes parts of the program doing actual computations.

+ Added the C 'string.h' header for the 'strlen' method as it is not always included in the C++ header.
+ Added macro for the path to the postscript viewer in 'iben.h', and inserted it instead of 'Ghostview'.
+ Changed the mentions of 'Ghostview' in the man page to postscript viewer.

Note: if you only want the patches and not the source code is a patch file provided in the patch folder.

##License
The program is licensed under version 2 of the GPL as IBEN originally was, and a copy of the license is bundled with the program. 


src/NEWS
February 27th, 2003, version 1.1
--------------------------------------

Released IBEN version 1.1 under GPL. This is the first GPL release.

Notice: 'make install' will not copy the README file. Without the
README file the help feature does not work.


src/iben.h
/*
 * IBEN, An Interactive BDD Environment
 * Copyright (C) 2003 Gerd Behrmann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef __IBEN_H_
#define __IBEN_H_

#include <map>
#include <string>

#include <bdd.h>
#include <string.h>

#define PSV "gsview"

#define EXPR 1
#define VARASSOC 2
#define NEW 3
#define VAR 4

struct st_entry
{
    int type;
    std::string id;
    struct {
	struct
	{
	    bdd f;
	} expr;
	struct
	{
	    bddPair *p;
	} varassoc;
	int var;
    } info;
  ~st_entry();
  void free();
};

extern bool terminated;
extern bool interactive;
extern bool timer_running;
extern int line_number;
extern std::map<std::string,st_entry> symbol_table;
extern std::map<int,std::string> name_table;
extern std::map<std::string,int> keywords;
extern double user_start;
extern double system_start;
extern double user_last;
extern double system_last;

extern void printdot(FILE*, bdd);
extern void print_sat(bdd);
extern const char *undefined(const char *);

extern void get_cpu_time(double *, double *);
extern int yyparse();
extern void parse(const char *);
extern void yyerror(const char *);
extern void print_version();

#endif


src/grammar.h
#ifndef BISON_GRAMMAR_H
# define BISON_GRAMMAR_H

#ifndef YYSTYPE
typedef union
{
    st_entry *e;
    bdd *f;
    long i;
    bddPair *p;
    char *s;
} yystype;
# define YYSTYPE yystype
# define YYSTYPE_IS_TRIVIAL 1
#endif
# define	T_UNKNOWNCHAR	257
# define	T_ID	258
# define	T_NEWID	259
# define	T_NUM	260
# define	T_CLEAR	261
# define	T_ALL	262
# define	T_LIMIT	263
# define	T_PRINT	264
# define	T_SHOW	265
# define	T_PRINTSOP	266
# define	T_SATISFY	267
# define	T_SATCOUNT	268
# define	T_SIZE	269
# define	T_STATS	270
# define	T_TIMER	271
# define	T_OFF	272
# define	T_VARS	273
# define	T_AGETS	274
# define	T_GETS	275
# define	T_EXISTS	276
# define	T_FORALL	277
# define	T_SUBST	278
# define	T_QUIT	279
# define	T_HELP	280
# define	T_IMPLIES	281
# define	T_NONE	282
# define	T_VERSION	283


extern YYSTYPE yylval;

#endif /* not BISON_GRAMMAR_H */


src/main.cc
/*
 * IBEN, An Interactive BDD Environment
 * Copyright (C) 2003 Gerd Behrmann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <cstdio>
#include <cstdlib>
#include <set>
#include <ctime>
#include "config.h"

#include <sys/resource.h>

extern "C" {
#ifdef HAVE_LIBREADLINE
#  if defined(HAVE_READLINE_READLINE_H)
#    include <readline/readline.h>
#  elif defined(HAVE_READLINE_H)
#    include <readline.h>
#  else /* !defined(HAVE_READLINE_H) */
extern char *readline ();
#  endif /* !defined(HAVE_READLINE_H) */
char *cmdline = NULL;
#else /* !defined(HAVE_READLINE_READLINE_H) */
#error "IBEN requires readline to compile"
#endif /* HAVE_LIBREADLINE */

#ifdef HAVE_READLINE_HISTORY
#  if defined(HAVE_READLINE_HISTORY_H)
#    include <readline/history.h>
#  elif defined(HAVE_HISTORY_H)
#    include <history.h>
#  else /* !defined(HAVE_HISTORY_H) */
extern void add_history ();
extern int write_history ();
extern int read_history ();
#  endif /* defined(HAVE_READLINE_HISTORY_H) */
  /* no history */
#endif /* HAVE_READLINE_HISTORY */
}

#ifdef NEW
#undef NEW
#endif
#include "iben.h"
#include "grammar.h"

using std::map;
using std::string;
using std::set;

extern int getrusage(int, struct rusage *);

bool terminated = false;
bool interactive = true;
bool timer_running = false;
int line_number;
map<string,st_entry> symbol_table;
map<int,string> name_table;
map<string,int> keywords;
double user_start;
double system_start;
double user_last;
double system_last;

const char *undefined(const char *id)
{
    static char buff[1000];
    snprintf(buff, 1000, "`%s' undefined", id);
    return buff;
}

static void make_keyword(const char *word, int code)
{
  keywords[word] = code;
}

void init_tables()
{
    make_keyword("clear", T_CLEAR);
    make_keyword("all", T_ALL);
    make_keyword("help", T_HELP);
    make_keyword("limit", T_LIMIT);
    make_keyword("print", T_PRINT);
    make_keyword("show", T_SHOW);
    make_keyword("none", T_NONE);
    make_keyword("satisfy", T_SATISFY);
    make_keyword("size", T_SIZE);
    make_keyword("sop", T_PRINTSOP);
    make_keyword("stats", T_STATS);
    make_keyword("timer", T_TIMER);
    make_keyword("off", T_OFF);
    make_keyword("vars", T_VARS);
    make_keyword("exists", T_EXISTS);
    make_keyword("forall", T_FORALL);
    make_keyword("subst", T_SUBST);
    make_keyword("quit", T_QUIT);
    make_keyword("satcount", T_SATCOUNT);
    make_keyword("version", T_VERSION);
}

st_entry::~st_entry()
{
    free();
}

void st_entry::free()
{
    switch (type)
    {
    case EXPR:
    case VAR:
      info.expr.f = bddfalse;
      break;
    case VARASSOC:
      bdd_freepair(info.varassoc.p);
      break;
    }
    type = NEW;
}

void get_cpu_time(double *user, double *system)
{
    struct rusage resources;

    getrusage(RUSAGE_SELF, &resources);
    *user=resources.ru_utime.tv_sec+resources.ru_utime.tv_usec/1.0e6;
    *system=resources.ru_stime.tv_sec+resources.ru_stime.tv_usec/1.0e6;
}

static const char *getname(int var)
{
  return name_table[var].c_str();
}

void print_sat(bdd r)
{
  bdd sat = bdd_satone(r);
  bdd_printset(r);
  printf("\n");
}

static set<int> visited;

static void printdot_rec(FILE* ofile, bdd b)
{
    if (b == bdd_false() || b == bdd_true()
	|| visited.find(b.id()) != visited.end())
    {
	return;
    }

    fprintf(ofile, "%d [label=\"%s\"];\n",
	    b.id(), getname(bdd_var(b)));
    if (bdd_low(b) != bdd_false())
	fprintf(ofile, "%d -> %d [style=dotted];\n", b.id(), bdd_low(b).id());
    if (bdd_high(b) != bdd_false())
	fprintf(ofile, "%d -> %d [style=filled];\n", b.id(), bdd_high(b).id());

    visited.insert(b.id());
   
    printdot_rec(ofile, bdd_low(b));
    printdot_rec(ofile, bdd_high(b));
}

void printdot(FILE* ofile, bdd r)
{
    visited.clear();
    fprintf(ofile, "digraph G {\n");
    fprintf(ofile, "preroot[style=invis]\n");
    if (r != bdd_false()) fprintf(ofile, "preroot->%d\n", r.id());
    fprintf(ofile, "1 [shape=box, label=\"1\", style=filled, shape=box, height=0.3, width=0.3];\n");

    printdot_rec(ofile, r);

    fprintf(ofile, "}\n");
}

static void print_copyright()
{
    printf("IBEN version 1.1, Copyright (C) 1998,2003 Gerd Behrmann\n");
    printf("IBEN comes with ABSOLUTELY NO WARRANTY\n");
    printf("This is free software, and you are welcome to redistribute it\n");
    printf("under the conditions of the GPL.\n\n");
}

void print_version()
{
  printf("IBEN version 1.1, Copyright (C) 1998,2003 Gerd Behrmann\n");
  printf("Compiled %s %s", __DATE__, __TIME__);
#ifdef __GNUC__    
  printf(" with GCC %s", __VERSION__);
#endif
  printf("\n\n");
}

int main(int argc, char **argv)
{
  char *s;
  print_copyright();
    
    bdd_init(100000, 20000);
    rl_bind_key ('\t', rl_insert);
    init_tables();
    line_number = 1;

    terminated = false;
    while (!terminated) {
	s = readline("iben> ");
	if (s == NULL)
	  break;
	if (*s) {
#ifdef HAVE_READLINE_HISTORY
	  add_history(s);
#endif
	  parse(s);
	}
	free(s);
    }
    printf("\n");
    return (0);
}
